File: crates/kuben-agent/src/hub.rs
//! The hub's end of AgentLink, hosted by `kuben serve` (M1.9 integration)
//! and tested end to end in this crate.
//!
//! * An anonymous peer may only enroll ([`serve_enrollment`]); the issued
//!   certificate is recorded in the [`Registry`].
//! * An authenticated peer is the cluster its certificate names (the URI
//!   SAN `kuben://cluster/<id>` the hub issued). Its Hello must name the same
//!   cluster, and its device must be the cluster's current, unrevoked agent
//!   device ([`Registry::admits`]): a revoked device, or one a newer
//!   enrollment replaced, is refused even while its certificate is valid.
//! * The handshake negotiates the protocol version and the features
//!   ([`negotiate`]); the hub then answers heartbeats and records the link
//!   and when it last heard from each cluster.
//! * A linked agent renews its certificate over the link, for the device
//!   key the link authenticated with ([`Enrollment::renew`]).

use std::{
    collections::{BTreeSet, HashMap},
    ops::RangeInclusive,
    sync::{Mutex, MutexGuard, PoisonError},
    time::{Duration, Instant},
};

use rustls::pki_types::CertificateDer;
use time::OffsetDateTime;
use tokio::io::{AsyncRead, AsyncWrite};
use tokio_rustls::server::TlsStream;
use x509_parser::prelude::{FromDer, GeneralName, ParsedExtension, X509Certificate};

use crate::{
    enroll::{Enrollment, TokenStore, device_id_of, serve_enrollment},
    protocol::{FrameError, Message, Refusal, SUPPORTED_VERSIONS, negotiate, read_frame, write_frame},
    tls::peer_certificate,
};

/// What the hub offers.
#[derive(Clone, Debug)]
pub struct HubSettings {
    pub hub_version: String,
    pub versions: RangeInclusive<u32>,
    pub features: BTreeSet<String>,
    /// How often agents send a heartbeat.
    pub heartbeat: Duration,
}

impl Default for HubSettings {
    fn default() -> Self {
        Self {
            hub_version: env!("CARGO_PKG_VERSION").to_owned(),
            versions: SUPPORTED_VERSIONS,
            features: BTreeSet::new(),
            heartbeat: Duration::from_secs(10),
        }
    }
}

/// What the hub knows of a cluster's latest link.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SessionInfo {
    pub version: u32,
    pub features: BTreeSet<String>,
    pub agent_version: String,
    pub heartbeats: u64,
    pub last_seen: Instant,
}

/// Where the hub keeps what it knows of each cluster's agent: SQL in `kuben
/// serve`, memory in tests. A registry that cannot answer refuses.
pub trait Registry: Send + Sync {
    /// Whether `device` is the current, unrevoked agent device of `cluster`.
    fn admits(&self, cluster: &str, device: &str) -> impl Future<Output = bool> + Send;

    /// A certificate valid until `not_after` was issued to `device` of
    /// `cluster`: it is the cluster's agent device from now on.
    fn certified(
        &self,
        cluster: &str,
        device: &str,
        not_after: OffsetDateTime,
    ) -> impl Future<Output = ()> + Send;

    /// `device` of `cluster` linked with `session`.
    fn linked(&self, cluster: &str, device: &str, session: &SessionInfo) -> impl Future<Output = ()> + Send;

    /// A heartbeat of `device` of `cluster` arrived.
    fn heard(&self, cluster: &str, device: &str) -> impl Future<Output = ()> + Send;
}

fn lock<V>(m: &Mutex<V>) -> MutexGuard<'_, V> {
    m.lock().unwrap_or_else(PoisonError::into_inner)
}

/// A registry in memory, for tests.
#[derive(Debug, Default)]
pub struct MemoryRegistry {
    /// Cluster → its agent device and whether that device is revoked.
    devices: Mutex<HashMap<String, (String, bool)>>,
}

impl MemoryRegistry {
    /// Revoke the current device of `cluster`.
    pub fn revoke(&self, cluster: &str) {
        if let Some(entry) = lock(&self.devices).get_mut(cluster) {
            entry.1 = true;
        }
    }
}

impl Registry for MemoryRegistry {
    fn admits(&self, cluster: &str, device: &str) -> impl Future<Output = bool> + Send {
        let admitted = lock(&self.devices)
            .get(cluster)
            .is_some_and(|(current, revoked)| current == device && !revoked);
        std::future::ready(admitted)
    }

    fn certified(
        &self,
        cluster: &str,
        device: &str,
        _not_after: OffsetDateTime,
    ) -> impl Future<Output = ()> + Send {
        let mut devices = lock(&self.devices);
        // The same device keeps its revocation; another one replaces it.
        if devices.get(cluster).is_none_or(|(current, _)| current != device) {
            devices.insert(cluster.to_owned(), (device.to_owned(), false));
        }
        std::future::ready(())
    }

    fn linked(
        &self,
        _cluster: &str,
        _device: &str,
        _session: &SessionInfo,
    ) -> impl Future<Output = ()> + Send {
        std::future::ready(())
    }

    fn heard(&self, _cluster: &str, _device: &str) -> impl Future<Output = ()> + Send {
        std::future::ready(())
    }
}

/// The hub's AgentLink endpoint.
#[derive(Debug)]
pub struct Hub<T, R> {
    enrollment: Enrollment<T>,
    registry: R,
    settings: HubSettings,
    sessions: Mutex<HashMap<String, SessionInfo>>,
}

/// The cluster a certificate the hub issued names.
#[must_use]
pub fn cluster_of(certificate: &CertificateDer<'_>) -> Option<String> {
    let (_, cert) = X509Certificate::from_der(certificate).ok()?;
    cert.extensions()
        .iter()
        .find_map(|ext| match ext.parsed_extension() {
            ParsedExtension::SubjectAlternativeName(san) => {
                san.general_names.iter().find_map(|name| match name {
                    GeneralName::URI(uri) => uri.strip_prefix("kuben://cluster/").map(str::to_owned),
                    _ => None,
                })
            }
            _ => None,
        })
}

/// The device id of the key a certificate is for.
fn device_of(certificate: &CertificateDer<'_>) -> Option<String> {
    let (_, cert) = X509Certificate::from_der(certificate).ok()?;
    Some(device_id_of(cert.public_key().raw))
}

async fn refuse<S: AsyncWrite + Unpin>(
    stream: &mut S,
    reason: Refusal,
    message: &str,
) -> Result<(), FrameError> {
    write_frame(
        stream,
        &Message::Refused {
            reason,
            message: message.to_owned(),
        },
    )
    .await
}

const NOT_THE_AGENT: &str =
    "this device is not the cluster's agent: revoked, or replaced by a newer enrollment";

impl<T: TokenStore, R: Registry> Hub<T, R> {
    #[must_use]
    pub fn new(enrollment: Enrollment<T>, registry: R, settings: HubSettings) -> Self {
        Self {
            enrollment,
            registry,
            settings,
            sessions: Mutex::new(HashMap::new()),
        }
    }

    #[must_use]
    pub const fn enrollment(&self) -> &Enrollment<T> {
        &self.enrollment
    }

    #[must_use]
    pub const fn registry(&self) -> &R {
        &self.registry
    }

    /// The latest link of `cluster` this hub served, if any.
    #[must_use]
    pub fn session(&self, cluster: &str) -> Option<SessionInfo> {
        lock(&self.sessions).get(cluster).cloned()
    }

    /// Serve one accepted TLS connection until it ends.
    pub async fn serve<S>(&self, mut tls: TlsStream<S>) -> Result<(), FrameError>
    where
        S: AsyncRead + AsyncWrite + Unpin,
    {
        let Some(certificate) = peer_certificate(tls.get_ref().1) else {
            let now = OffsetDateTime::now_utc();
            if let Some(issued) = serve_enrollment(&mut tls, &self.enrollment, now).await? {
                self.registry
                    .certified(&issued.cluster_id, &issued.device_id, issued.not_after)
                    .await;
            }
            return Ok(());
        };
        let (Some(cluster), Some(device)) = (cluster_of(&certificate), device_of(&certificate)) else {
            return refuse(
                &mut tls,
                Refusal::UnknownCluster,
                "the certificate names no cluster",
            )
            .await;
        };
        let Some(Message::Hello {
            protocol_versions,
            agent_version,
            cluster_id,
            capabilities,
            ..
        }) = read_frame(&mut tls).await?
        else {
            return refuse(&mut tls, Refusal::BadRequest, "a link opens with Hello").await;
        };
        if cluster_id != cluster {
            return refuse(
                &mut tls,
                Refusal::UnknownCluster,
                "Hello names another cluster than the certificate",
            )
            .await;
        }
        if !self.registry.admits(&cluster, &device).await {
            return refuse(&mut tls, Refusal::Revoked, NOT_THE_AGENT).await;
        }
        let agreed = match negotiate(
            &self.settings.versions,
            &self.settings.features,
            &protocol_versions,
            &capabilities,
        ) {
            Ok(agreed) => agreed,
            Err(reason) => return refuse(&mut tls, reason, "no protocol version in common").await,
        };
        write_frame(
            &mut tls,
            &Message::Welcome {
                protocol_version: agreed.version,
                hub_version: self.settings.hub_version.clone(),
                heartbeat_ms: u64::try_from(self.settings.heartbeat.as_millis()).unwrap_or(u64::MAX),
                features: agreed.features.clone(),
            },
        )
        .await?;
        let session = SessionInfo {
            version: agreed.version,
            features: agreed.features,
            agent_version,
            heartbeats: 0,
            last_seen: Instant::now(),
        };
        self.registry.linked(&cluster, &device, &session).await;
        lock(&self.sessions).insert(cluster.clone(), session);
        self.converse(&mut tls, &cluster, &device).await
    }

    /// The linked part of a connection: heartbeats and renewals until the
    /// agent hangs up.
    async fn converse<S>(&self, tls: &mut TlsStream<S>, cluster: &str, device: &str) -> Result<(), FrameError>
    where
        S: AsyncRead + AsyncWrite + Unpin,
    {
        loop {
            match read_frame(tls).await? {
                Some(Message::Heartbeat { seq }) => {
                    if let Some(session) = lock(&self.sessions).get_mut(cluster) {
                        session.heartbeats += 1;
                        session.last_seen = Instant::now();
                    }
                    self.registry.heard(cluster, device).await;
                    write_frame(tls, &Message::HeartbeatAck { seq }).await?;
                }
                Some(Message::Renew { csr }) => {
                    if !self.registry.admits(cluster, device).await {
                        return refuse(tls, Refusal::Revoked, NOT_THE_AGENT).await;
                    }
                    let issued = match self
                        .enrollment
                        .renew(cluster, &csr, device, OffsetDateTime::now_utc())
                    {
                        Ok(issued) => issued,
                        Err(reason) => {
                            return refuse(tls, reason, "a renewal is for the device key of this link").await;
                        }
                    };
                    self.registry.certified(cluster, device, issued.not_after).await;
                    write_frame(
                        tls,
                        &Message::Enrolled {
                            certificate: issued.certificate_pem,
                            not_after: issued.not_after.unix_timestamp(),
                        },
                    )
                    .await?;
                }
                Some(Message::Unknown) => {}
                Some(_) => return refuse(tls, Refusal::BadRequest, "unexpected message").await,
                None => return Ok(()),
            }
        }
    }
}

File: apps/kuben/internal/agentlink/hub.go
package agentlink

// The hub's end of AgentLink (hub.rs), hosted by `kuben serve`:
//
//   - An anonymous peer may only enroll ([ReceiveEnrollment]); the device is
//     recorded in the [Registry] before [AnswerEnrollment] hands the
//     certificate over.
//   - An authenticated peer is the cluster its certificate names (the URI
//     SAN `kuben://cluster/<id>` the hub issued). Its Hello must name the
//     same cluster, and its device must be the cluster's current, unrevoked
//     agent device ([Registry.Admits]): a revoked device, or one a newer
//     enrollment replaced, is refused even while its certificate is valid.
//   - The handshake negotiates the protocol version and the features; the
//     hub then answers heartbeats and records the link and when it last
//     heard from each cluster.
//   - A linked agent renews its certificate over the link, for the device
//     key the link authenticated with ([Enrollment.Renew]).
//   - [Hub.Send] hands a linked agent an execution envelope, when both sides
//     negotiated [protocol.ApplicationRuntime]; the agent's observations go
//     to [Registry.Observed].

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"errors"
	"math"
	"strings"
	"sync"
	"time"

	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/clock"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/version"
	protocol "github.com/Teamtem-dev/kuben/packages/api/agentlink"
)

// HubSettings is what the hub offers.
type HubSettings struct {
	HubVersion string
	Versions   protocol.Versions
	Features   protocol.Features
	// Heartbeat is how often agents send a heartbeat.
	Heartbeat time.Duration
}

// DefaultHubSettings is this build's version and protocol versions, no
// features, a heartbeat every 10 seconds.
func DefaultHubSettings() HubSettings {
	return HubSettings{
		HubVersion: version.Version,
		Versions:   protocol.SupportedVersions,
		Features:   protocol.Features{},
		Heartbeat:  10 * time.Second,
	}
}

// SessionInfo is what the hub knows of a cluster's latest link.
type SessionInfo struct {
	Version      uint32
	Features     protocol.Features
	AgentVersion string
	Heartbeats   uint64
	// LastSeen is unix milliseconds.
	LastSeen int64
}

// Registry is where the hub keeps what it knows of each cluster's agent:
// SQL in `kuben serve`, memory in tests. A registry that cannot answer
// refuses.
type Registry interface {
	// Admits says whether device is the current, unrevoked agent device of
	// cluster.
	Admits(ctx context.Context, cluster, device string) bool
	// Certified: a certificate valid until notAfter was issued to device of
	// cluster; it is the cluster's agent device from now on.
	Certified(ctx context.Context, cluster, device string, notAfter time.Time)
	// Linked: device of cluster linked with session.
	Linked(ctx context.Context, cluster, device string, session SessionInfo)
	// Heard: a heartbeat of device of cluster arrived.
	Heard(ctx context.Context, cluster, device string)
	// Observed: device of cluster reported observation of an envelope.
	Observed(ctx context.Context, cluster, device string, observation protocol.Observation)
}

// linked is a cluster's latest link: what it negotiated, and the way to
// send it envelopes.
type linked struct {
	info   SessionInfo
	outbox chan protocol.Message
	// done is closed when the link ended.
	done chan struct{}
}

// Hub is the hub's AgentLink endpoint.
type Hub struct {
	enrollment *Enrollment
	registry   Registry
	settings   HubSettings
	clock      clock.Clock

	mu sync.Mutex
	// sessions are the latest link of each cluster, guarded by mu.
	sessions map[string]*linked
}

// NewHub is a hub issuing with enrollment and recording in registry.
func NewHub(enrollment *Enrollment, registry Registry, settings HubSettings, clk clock.Clock) *Hub {
	return &Hub{enrollment: enrollment, registry: registry, settings: settings, clock: clk, sessions: map[string]*linked{}}
}

// Enrollment is the hub's enrollment service.
func (h *Hub) Enrollment() *Enrollment { return h.enrollment }

// Registry is where the hub records its agents.
func (h *Hub) Registry() Registry { return h.registry }

func (h *Hub) now() time.Time { return time.UnixMilli(h.clock.NowMs()) }

// Session is the latest link of cluster this hub served, if any.
func (h *Hub) Session(cluster string) (SessionInfo, bool) {
	h.mu.Lock()
	defer h.mu.Unlock()
	l := h.sessions[cluster]
	if l == nil {
		return SessionInfo{}, false
	}
	info := l.info
	info.Features = append(protocol.Features{}, l.info.Features...)
	return info, true
}

// Send hands the agent of cluster the envelope apply over its live link.
// False when it is not linked, or did not negotiate the runtime feature.
func (h *Hub) Send(ctx context.Context, cluster string, apply protocol.Apply) bool {
	h.mu.Lock()
	l := h.sessions[cluster]
	usable := l != nil && l.info.Features.Has(protocol.ApplicationRuntime)
	h.mu.Unlock()
	if !usable || l == nil {
		return false
	}
	select {
	case <-l.done:
		return false
	default:
	}
	select {
	case l.outbox <- apply:
		return true
	case <-l.done:
		return false
	case <-ctx.Done():
		return false
	}
}

// ClusterOf is the cluster a certificate the hub issued names.
func ClusterOf(cert *x509.Certificate) (string, bool) {
	for _, uri := range cert.URIs {
		if cluster, ok := strings.CutPrefix(uri.String(), "kuben://cluster/"); ok {
			return cluster, true
		}
	}
	return "", false
}

const notTheAgent = "this device is not the cluster's agent: revoked, or replaced by a newer enrollment"

func refuse(conn *tls.Conn, reason protocol.Refusal, message string) error {
	return protocol.WriteFrame(conn, protocol.Refused{Reason: reason, Message: message})
}

// Serve serves one accepted TLS connection until it ends or ctx does, and
// closes it.
func (h *Hub) Serve(ctx context.Context, conn *tls.Conn) error {
	defer conn.Close() //nolint:errcheck // the link is over either way
	if err := conn.HandshakeContext(ctx); err != nil {
		return err //nolint:wrapcheck // crypto/tls says what failed
	}
	cert, authenticated := protocol.PeerCertificate(conn.ConnectionState())
	if !authenticated {
		issued, ok, err := ReceiveEnrollment(ctx, conn, h.enrollment, h.now())
		if err != nil || !ok {
			return err
		}
		// Recorded before the agent has its certificate: it links at once,
		// and a Hello the registry does not know yet would be refused as
		// revoked.
		h.registry.Certified(ctx, issued.ClusterID, issued.DeviceID, issued.NotAfter)
		return AnswerEnrollment(conn, issued)
	}
	cluster, named := ClusterOf(cert)
	if !named {
		return refuse(conn, protocol.RefusalUnknownCluster, "the certificate names no cluster")
	}
	device := protocol.DeviceIDOf(cert.RawSubjectPublicKeyInfo)
	m, _, err := protocol.ReadFrame(conn)
	if err != nil {
		return err //nolint:wrapcheck // a frame error says what failed
	}
	hello, isHello := m.(protocol.Hello)
	switch {
	case !isHello:
		return refuse(conn, protocol.RefusalBadRequest, "a link opens with Hello")
	case hello.ClusterID != cluster:
		return refuse(conn, protocol.RefusalUnknownCluster, "Hello names another cluster than the certificate")
	case !h.registry.Admits(ctx, cluster, device):
		return refuse(conn, protocol.RefusalRevoked, notTheAgent)
	}
	agreed, err := protocol.Negotiate(h.settings.Versions, h.settings.Features, hello.ProtocolVersions, hello.Capabilities)
	if err != nil {
		var reason protocol.Refusal
		if !errors.As(err, &reason) {
			reason = protocol.RefusalUnsupportedProtocol
		}
		return refuse(conn, reason, "no protocol version in common")
	}
	welcome := protocol.Welcome{
		ProtocolVersion: agreed.Version,
		HubVersion:      h.settings.HubVersion,
		HeartbeatMs:     uint64(min(h.settings.Heartbeat.Milliseconds(), math.MaxInt64)), //nolint:gosec // a duration is not negative
		Features:        agreed.Features,
	}
	if err := protocol.WriteFrame(conn, welcome); err != nil {
		return err //nolint:wrapcheck // a frame error says what failed
	}
	session := SessionInfo{Version: agreed.Version, Features: agreed.Features, AgentVersion: hello.AgentVersion, LastSeen: h.clock.NowMs()}
	h.registry.Linked(ctx, cluster, device, session)
	l := &linked{info: session, outbox: make(chan protocol.Message, 16), done: make(chan struct{})}
	h.mu.Lock()
	h.sessions[cluster] = l
	h.mu.Unlock()
	defer close(l.done)
	return h.converse(ctx, conn, cluster, device, l)
}

// frame is one read of the link.
type frame struct {
	message protocol.Message
	open    bool
	err     error
}

// readFrames reads conn until it fails or ends, handing each read to
// frames until stop closes.
func readFrames(conn *tls.Conn, frames chan<- frame, stop <-chan struct{}) {
	for {
		m, open, err := protocol.ReadFrame(conn)
		select {
		case frames <- frame{message: m, open: open, err: err}:
		case <-stop:
			return
		}
		if err != nil || !open {
			return
		}
	}
}

// converse is the linked part of a connection until the agent hangs up.
// Frames are read by a goroutine of their own, so none is lost halfway
// while an envelope goes out.
func (h *Hub) converse(ctx context.Context, conn *tls.Conn, cluster, device string, l *linked) error {
	frames := make(chan frame, 16)
	stop := make(chan struct{})
	var reading sync.WaitGroup
	reading.Go(func() { readFrames(conn, frames, stop) })
	defer func() {
		close(stop)
		conn.Close() //nolint:errcheck,gosec // unblocks the reader; the link is over
		reading.Wait()
	}()
	for {
		select {
		case <-ctx.Done():
			return nil
		case command := <-l.outbox:
			if err := protocol.WriteFrame(conn, command); err != nil {
				return err //nolint:wrapcheck // a frame error says what failed
			}
		case f := <-frames:
			switch {
			case f.err != nil:
				return f.err
			case !f.open:
				return nil
			}
			goOn, err := h.handle(ctx, conn, f.message, cluster, device)
			if err != nil || !goOn {
				return err
			}
		}
	}
}

// handle is one message of a linked agent; false when the link must
// close.
func (h *Hub) handle(ctx context.Context, conn *tls.Conn, m protocol.Message, cluster, device string) (bool, error) {
	switch m := m.(type) {
	case protocol.Heartbeat:
		h.mu.Lock()
		if l := h.sessions[cluster]; l != nil {
			l.info.Heartbeats++
			l.info.LastSeen = h.clock.NowMs()
		}
		h.mu.Unlock()
		h.registry.Heard(ctx, cluster, device)
		return true, protocol.WriteFrame(conn, protocol.HeartbeatAck(m))
	case protocol.Renew:
		if !h.registry.Admits(ctx, cluster, device) {
			return false, refuse(conn, protocol.RefusalRevoked, notTheAgent)
		}
		issued, err := h.enrollment.Renew(cluster, m.CSR, device, h.now())
		if err != nil {
			reason := protocol.RefusalBadRequest
			if r, ok := errors.AsType[protocol.Refusal](err); ok {
				reason = r
			}
			return false, refuse(conn, reason, "a renewal is for the device key of this link")
		}
		h.registry.Certified(ctx, cluster, device, issued.NotAfter)
		return true, protocol.WriteFrame(conn, protocol.Enrolled{
			Certificate: issued.CertificatePEM, NotAfter: issued.NotAfter.Unix(),
		})
	case protocol.Observation:
		h.registry.Observed(ctx, cluster, device, m)
		return true, nil
	case protocol.Unknown:
		return true, nil
	case protocol.Hello, protocol.Welcome, protocol.Refused, protocol.HeartbeatAck, protocol.Enroll,
		protocol.Enrolled, protocol.Apply:
	}
	return false, refuse(conn, protocol.RefusalBadRequest, "unexpected message")
}
